lab_4_retrieval_w_clustering/main.py

```diff
@@ -314,8 +483,29 @@ def get_clusters_info(self, num_examples: int) -> list[dict[str, int | list[str]
             num_examples (int): Number of examples for each cluster
 
         Returns:
-            list[dict[str, int| list[str]]]: List with information about each cluster
-        """
+            list[dict[str, int | list[str]]]: List with information about each cluster
+        """
+        if not (isinstance(num_examples, int) and num_examples > 0):
+            raise ValueError("The argument is not a natural number")
+        list_of_cluster_info = []
+        for cluster_index, cluster in enumerate(self.__clusters):
+            cluster_vectors = self._db.get_vectors(cluster.get_indices())
+            vector_distances = []
+            for index, vector in cluster_vectors:
+                vector_distance = calculate_distance(cluster.get_centroid(), vector)
+                if not isinstance(vector_distance, float):
+                    raise ValueError("Failed to calculate the distance to the centroid")
+                vector_distances.append((index, vector_distance))
+            vector_distances = sorted(vector_distances, key=lambda a: a[1])[:num_examples]
+            doc_indices = tuple(pair[0] for pair in vector_distances)
+            documents = self._db.get_raw_documents(doc_indices)
+            if (isinstance(cluster_index, int) and isinstance(documents, list) and
+                all(isinstance(document, str) for document in documents)):
+                info_dict = {"cluster_id": cluster_index, "documents": documents}
+                list_of_cluster_info.append(info_dict)
+            else:
+                raise ValueError("Failed to get raw documents")
+        return list_of_cluster_info
```

Comment:
to cope with mypy, try first saving all cluster ids and documents as keys and values ​​in a dictionary, and only after finishing all the loops, make a report

Reply:
Cannot split by "\n", the put_corpus method fails

Reply:
Slightly altered the get_paragraphs method

Reply:
MyPy is finally calm now, thank you!